### ISSUE TITLE
deprecate getProgramsMenuPath and getProgramFilesPath in win32.py

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#7883]
|<img alt="opalmer's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| opalmer reported|
|-|-|
|Trac ID|trac#7883|
|Type|enhancement|
|Created|2015-05-03 18:54:25Z|
|Branch|https://github.com/twisted/twisted/tree/win32-deprecate-7883-2|

In an effort to reduce Twisted's dependency on win32api some calls, such as those calling win32api.Reg*, can be switched to Python's built-in registry modules (_winreg in Python 2 or winreg in Python 3).    This could be a part of the work put into [#7477](https://github.com/twisted/twisted/issues/7477) but it's probably safer and more reliable to use Python's built-in modules instead of building our own with cffi.

Attachments:

* [7883.v1.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/5e4/5e443c003151351b220521f436b12a2ccdd82f6e/f5060824ddd9bc86efdcd5f3dca714b2071127b0.patch) (7110 bytes) - added by opalmer on 2015-05-03 19:33:51Z - 
* [7883.v2.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/5e4/5e443c003151351b220521f436b12a2ccdd82f6e/45fa162d71fad9f957b3f9e1c2bcb3e67d360dea.patch) (7117 bytes) - added by opalmer on 2015-05-03 19:37:51Z - don't reraise ImportError in compact (it will break other platforms)
* [7883.v3.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/5e4/5e443c003151351b220521f436b12a2ccdd82f6e/feb7f0dc97b5880db242c114a2b5fea0d93004da.patch) (6702 bytes) - added by opalmer on 2015-05-03 21:25:31Z - 
* [7883.v4.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/5e4/5e443c003151351b220521f436b12a2ccdd82f6e/8d99da3ff58d47833865715e83878ccbebeadd98.patch) (1385 bytes) - added by opalmer on 2015-05-10 01:29:58Z - 
* [7883.v4.2.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/5e4/5e443c003151351b220521f436b12a2ccdd82f6e/610d1da92c0b65af3e8bd9222efbcef96a848ff4.patch) (1385 bytes) - added by opalmer on 2015-05-10 01:30:43Z - 
* [7883.v5.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/5e4/5e443c003151351b220521f436b12a2ccdd82f6e/863f18fbf932fc597f1ce26e0f36b263c4ce573a.patch) (1342 bytes) - added by opalmer on 2015-05-18 01:45:24Z - 
* [7883.v6.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/5e4/5e443c003151351b220521f436b12a2ccdd82f6e/0bc3f08a3d7b8bcda5b378302f130399495c4bc9.patch) (4251 bytes) - added by opalmer on 2015-05-23 18:00:49Z - 
* [7883.v7.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/5e4/5e443c003151351b220521f436b12a2ccdd82f6e/d9de737fe8323e99420913ba0944661a42e8939f.patch) (3316 bytes) - added by opalmer on 2015-05-24 16:17:39Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__7883 7883
type__enhancement enhancement
reporter__opalmer opalmer
priority__normal normal
milestone__ 
branch__branches_win32_deprecate_7883_2 branches/win32-deprecate-7883-2
branch_author__adiroiban adiroiban
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1430679265612245 1430679265612245
changetime__1435494983099695 1435494983099695
version__None None
owner__adiroiban adiroiban
cc__thijs
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_1"></a>|@glyph commented|
|-|-|

I had originally thought the module was called `_winreg` to discourage its use. But looking at the docs real quick I see it's been moved to `winreg` on py3 to make it public, and py2 is never going to change, so we can rely on it there. Is it available on pypy?

I'd be happy to review a patch which makes this change.

**Comment by twisted-trac** [comment]
|<img alt="opalmer's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_2"></a>|opalmer commented|
|-|-|

Replying to [glyph](#note_1):
> I had originally thought the module was called `_winreg` to discourage its use. But looking at the docs real quick I see it's been moved to `winreg` on py3 to make it public, and py2 is never going to change, so we can rely on it there. Is it available on pypy?
> 

According to http://pypy.org/compat.html pypy seems to support winreg out of the box.  They seem to be importable too in pypy and pypy3 under the same namespaces as the CPython equivalent as well.

> I'd be happy to review a patch which makes this change.

I've got a patch ready for a first look actually (attached).  I'm still diving into all the various style/contributing guides so it's possible I've missed something but I'm happy to fix it of course.  The tests are passing for me locally but I've yet to run the full test suite (though a quick grep seems to indicate Twisted itself does not use those functions).

**Comment by twisted-trac** [comment]
|<img alt="opalmer's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|opalmer removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_4"></a>|@glyph commented|
|-|-|

See the ReviewProcess wiki page if you have more questions about workflow and coding standards.

**Comment by twisted-trac** [comment]
|<img alt="opalmer's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_5"></a>|opalmer commented|
|-|-|

Thanks glyph.  Looked that over and made a couple of adjustments in v3 which removes the now unused win32con/win32api imports.  I also removed the changes to compact.py since python.win32 is the only module using winreg.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_6"></a>|@exarkun set owner to opalmer|
|-|-|

Thanks.  Here are some comments:

  1. First, most broadly, these functions are used nowhere in Twisted.  As I recall, they were support APIs for a previous generation of Twisted's Windows installer generator.  Since that is gone, it's possible Twisted should stop providing these APIs (after a deprecation period).  That's probably the easiest way to remove this pywin32 dependency.  If so, I suggest submitting a patch that deprecates these functions rather than changing their implementations/tests.
  1. I think "Python 3" is probably sufficient.  I'm not sure what "Python 3+" might refer to. :)
  1. `NotImplemented` is a special value for operator implementations.  Just use `None` to signal there's no `winreg` module at all.
    1. As a bonus (ie, optional to address this): it's probably pretty easy to predict on which platforms winreg is available and on which it's not.  If the test suite is running on Windows and you fail to get the module, something is probably wrong and maybe some tests should fail - not skip.
  1. The class docstring for `FakeOpenKey` is formatted incorrectly.  Place the triple quotes on their own lines.
  1. The class docstring for `TestException` is also misformatted in the same way.
  1. Instead of monkey-patching the `winreg` module, consider making the module a parameter to the code being tested.  Or make `OpenKey`, `QueryValueEx`, etc parameters.  You can do this by making them arguments accepted by the functions being tested, for example.  It's easier to do this sort of thing when there's an instance, though - since `__init__` is a nicer place to put this sort of thing (including default values) than in each separate function that needs to be parameterized.
  1. Twisted naming convention is `camelCase`, not `under_scores`.  Various names in the patch need to be adjusted - `raise_exception`, `open_key`, `test_getProgramsMenuPath_closes_key` (which is tricky - it should be `test_getProgramsMenuPathClosesKey` - though an alternate, somewhat cleaner solution is to give `getProgramsMenuPath` a whole test case to itself so you can name its test methods like `test_closesKey`).
  1. Some test methods are missing docstrings.  They should all have docstrings which describe the intent of the test (so a future maintainer can possibly debug them when they fail, fix bugs discovered in them, or understand how it is reasonable to change the actual API being tested).
  1. Ideally, a test method asserts just one thing.  If there are multiple things to verify about an API, there should be multiple tests.
  1. What is the type of the key name accepted by `OpenKey` and `QueryValueEx`?  Is it bytes or unicode or does it depend on whether Python 2 or Python 3 is being used?  I can't find this information in the API documentation.
  1. Ideally `FakeOpenKey` would be a verified fake of the Python HKEY wrapper type.  As written now, it is a fake that we just have to hope faithfully mirrors the HKEY API (both now and in the future, as Python changes).  A verified fake would actually prove it is a sufficiently accurate simulacrum of the HKEY wrapper - and provide a failing test if somehow this ceases to be.  For example, you might write a test that closes the key, tries to use it, and asserts that a certain exception is raised (I'm assuming some exception is raised by the real implementation if you use it after it is closed).  Then, run this one test against both the real and fake implementations.  If both tests pass, you know the fake is close enough to the real to use.
  1. `test_getProgramFilesPath` and `test_getProgramsMenuPath` look like they are mostly duplicates of the implementations of the functions they test now.  This is a form that brittleness takes in unit tests.  If the test is exactly the implementation, it primarily results in doubling the cost of maintenance on the API.  If you want to change something, you have to change both the implementation and the test in exactly the same way.  The only thing this kind of test guards against is accidental changes to the implementation.  Twisted's pre-commit review process is probably a better guard against such changes.  Is there a different way these APIs could be tested?  A way that produces a less brittle test?  I know that with the kind of system interactions going on in these APIs it's often difficult to avoid this problem.  I'm not familiar enough with the Windows registry APIs to make any specific suggestions, though.

Thanks again.

**Comment by twisted-trac** [comment]
|<img alt="opalmer's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_7"></a>|opalmer commented|
|-|-|

Thanks for the review, sorry about the naming conventions I could have saved you some time there...

On the note of deprecation, I'm in favor of this unless someone can come up with a reason not to deprecate those functions.  I'd see the implementation working one of two ways:

1. Add a deprecation message and leave the pywin32 imports alone, do not convert to winreg.  Since twisted does not use the deprecated functions twisted itself is not dependent on pywin32 in this case.  External users of these functions however would require pywin32 but there shouldn't be very many cases of that based on how your description of their original use.
2. Add a deprecation message and switch to winreg.  This ensures that anyone using these functions does not have to have pywin32 installed to use this module.

I'd favor the first option because at the moment twisted on Windows requires pywin32 (or pypiwin32) to be installed anyway.  There's also not any functional changes to implement option one either compared to option two which simplifies the patch quite a bit.  I also did a bit of digging on a couple of open source search engines and didn't find any references to the functions to be deprecated either, not that it means nobody is using them just that they're not in common use everywhere.

**Comment by twisted-trac** [comment]
|<img alt="opalmer's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|opalmer set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|<img alt="opalmer's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_10"></a>|opalmer removed owner<br>opalmer set status to `new`|
|-|-|

Please have a look at 7883.v4.patch​ (7883.v4.2.patch​ is a duplicate...).  I've added deprecation warnings to win32.getProgramsMenuPath and win32.getProgramFilesPath.  I was not sure of what version should be included in the deprecation warning so I've set it up with the next micro release number, 15.1.1.  I'm happy to change the deprecation version though if someone  has something else in mind (maybe 15.2.0?).

**Comment by twisted-trac** [comment]
|<img alt="adiroiban1's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_11"></a>|adiroiban1 commented|
|-|-|

Changes look good.


15.1 was already released and 15.2 is in pre-release ... this should be update to 15.3

Deprecation should be done only at major/minor release... as bugfix releases don't have any change in the API... so your deprecation should look like:

```
@deprecated(Version("Twisted", 15, 3)) 
```

Thanks!

**Comment by twisted-trac** [comment]
|<img alt="opalmer's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_12"></a>|opalmer commented|
|-|-|

Thanks for the review, patch v5 attached.

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_13"></a>|@adiroiban commented|
|-|-|

(In [44881]) Branching to win32-deprecate-7883.

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_14"></a>|@adiroiban set owner to opalmer|
|-|-|

Thanks for working on this.

Changes need tests. It is highly recommend that while working on Twisted you use a development technique similar to TDD. ie, write a test to check that methods are deprecated, see the test fail, update the  code, see the test pass.

Let me know if you need help with writing test for deprecation behaviour.
You can check current test for example of deprecation tests.

meanwhile I sent your current patch to the test builders.

Please try writing the missing test or ask for help.

Thanks!

**Comment by twisted-trac** [comment]
|<img alt="opalmer's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_15"></a>|opalmer set status to `assigned`|
|-|-|

Will do, I'll get the tests written this weekend and thanks for starting a build.

Didn't think to include tests initially because deprecation itself has tests and the new tests would mainly ensure the deprecation was not removed (though it would prove deprecation was working too).  No objections here though, just something I'll keep in mind in the future.  Thanks.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_16"></a>|@glyph commented|
|-|-|

We mainly want to ensure that the deprecation works, and also, we want to make sure that we have tests such that the deprecation doesn't cause a warning in _our_ test suite; our test suite should explicitly be swallowing all warnings that it's producing intentionally.  When we remove the deprecated functionality, we will remove the tests with it :).

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_17"></a>|@glyph commented|
|-|-|

Thanks again, opalmer, for your contribution!

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_18"></a>|@adiroiban commented|
|-|-|

Another reason for writing tests for deprecation is that any new code should be run before commit.

You have the option either to manually start python, set up warnings, import the methods and check that warning are raised...
but then the reviewer will need to run the same steps again.

... so why not automate all this process into a unit test ? :)

**Comment by twisted-trac** [comment]
|<img alt="opalmer's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_19"></a>|opalmer removed owner<br>opalmer set status to `new`|
|-|-|

Oh I definitely saw the reasoning behind wanting the tests, even for deprecation, it's just the first time I've seen a request to test the deprecation behavior itself.  Nothing wrong with that at all and if anything it's a great way to ensure that from a consumer perspective it's working exactly as intended rather than assuming it is (that kind of approach often leads to a lot of problems over time).

Anyway, I've uploaded 7883.v6.patch which includes tests.  I ended up writing four tests, two for each deprecated function, to test both that the doc string is updated with the deprecated version and to make sure that the deprecation warning is in fact emitted.  Let me know if there's something else I should be testing and I'll be happy to add it.

> Changed 2 days ago by glyph
> Thanks again, opalmer, for your contribution!

Anytime :).  Figured this was the easiest thing to touch on first in terms of removing the pywin32 dependency.  Plus it's been a great introduction to the review process, coding standards, etc.  I'll need to make some updates [#7889](https://github.com/twisted/twisted/issues/7889) to reflect some of what I've learned here too.

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_20"></a>|@adiroiban set owner to opalmer|
|-|-|

Many thanks for the changes!

Not sure if we need tests for the docstring deprecation.

the test should check that a single warning is emitted, ie check that length of warnings is 1.

emitted_events should be named emittedEvents based on Twisted coding standard.

see twistedchecker results: https://buildbot.twistedmatrix.com/builders/twistedchecker/builds/3028/steps/run-twistedchecker/logs/new%20twistedchecker%20errors


The twisted test suite already catches all warnings and you can use `self.flushWarnings` to get the emitted warnings.

here is an example of deprecation tests using flushWarnings

```
class SSLContextFactoryTests(unittest.TestCase):
    """
    Tests for twisted.mail.protocols.SSLContextFactory.
    """
    def test_deprecation(self):
        """
        Accessing L{twisted.mail.protocols.SSLContextFactory} emits a
        deprecation warning recommending the use of the more general SSL context
        factory from L{twisted.internet.ssl}.
        """
        mail.protocols.SSLContextFactory
        warningsShown = self.flushWarnings([self.test_deprecation])
        self.assertEqual(len(warningsShown), 1)
        self.assertIdentical(warningsShown[0]['category'], DeprecationWarning)
        self.assertEqual(
            warningsShown[0]['message'],
            'twisted.mail.protocols.SSLContextFactory was deprecated in '
            'Twisted 12.2.0: Use twisted.internet.ssl.'
            'DefaultOpenSSLContextFactory instead.')

```

Other than that, changes look good and test pass... with the exception of Twistedchecker

Please check my commend and submit a new patch. thanks!

**Comment by twisted-trac** [comment]
|<img alt="opalmer's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_21"></a>|opalmer removed owner|
|-|-|

Updated, thanks again for the feedback!

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_22"></a>|@adiroiban commented|
|-|-|

(In [44927]) Branching to win32-deprecate-7883-2.

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_23"></a>|@adiroiban set owner to opalmer|
|-|-|

Changes look good.

I have reduced the text from test docstrings. Words like 'Test to ensure..' or 'should' add no value to the test and this is why we write the test in the first place. See https://plus.google.com/+JonathanLange/posts/YA3ThKWhSAj

Please check my changes and let me know if you are happy with them... or send corrections :)

I will merge the branch is you are happy with it.

Changes sent to buildbot.

Thanks again!

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_24"></a>|@adiroiban set status to `closed`|
|-|-|

(In [44954]) Merge win32-deprecate-7883-2: Deprecate getProgramsMenuPath and getProgramFilesPath in win32.py.

Author: opalmer
Reviewers: exarkun, adiroiban
Fixes: [#7883](https://github.com/twisted/twisted/issues/7883)

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_25"></a>|@adiroiban set status to `reopened`|
|-|-|

(In [44962]) Revert r44954: Deprecate getProgramsMenuPath and getProgramFilesPath in win32.py."

Tests are broken due to an untested last minute refactoring.

Reopens: [#7883](https://github.com/twisted/twisted/issues/7883)

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_26"></a>|@adiroiban commented|
|-|-|

OK. Builds are green on latest commit.

They were green before doing the commit, but I have not checked the revision at which the test ran.

Sorry for the trouble and the noise.

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_27"></a>|@adiroiban removed owner<br>@adiroiban set status to `new`|
|-|-|

Putting this for review so that another developer can check it before merge.

Thanks!

**Comment by twisted-trac** [comment]
|<img alt="opalmer's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_28"></a>|opalmer commented|
|-|-|

@adiroiban the doc string changes you suggested look good to me thanks!  Sorry for the delayed response, had something else I needed to attend to for a couple of days.

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_29"></a>|@hawkowl set owner to @adiroiban|
|-|-|

Looks good to me. I respun the builders to ensure the new tests work on the ones added since this was put up for review, and I am sure they should be all green.

Adi, once the buildbots complete and are green, go ahead and merge.

Thanks Adi and opalmer! :)

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_30"></a>|@adiroiban set status to `closed`|
|-|-|

(In [45121]) Merge win32-deprecate-7883-2: Deprecate getProgramsMenuPath and getProgramFilesPath in win32.py.

Author: opalmer
Reviewers: exarkun, adiroiban, hawkowl
Fixes: [#7883](https://github.com/twisted/twisted/issues/7883)
